### PR TITLE
Fix ambiguity error on construction from `ChainedVector`

### LINKED ISSUE #361: Trouble with `categorical(v)` when `v` is a `SentinelArrays.ChainedVector`
```julia
using SentinelArrays, CategoricalArrays
v = ChainedVector([1:3, 4:5])
@assert v isa AbstractVector{Int}

julia> categorical(v)
ERROR: MethodError: copyto!(::CategoricalArrays.CategoricalVector{Int64, UInt32, Int64, CategoricalArrays.CategoricalValue{Int64, UInt32}, Union{}}, ::ChainedVector{Int64, UnitRange{Int64}}) is ambiguous. Candidates:
  copyto!(dest::Union{SubArray{var"#s5", N, var"#s4", I, L} where {var"#s5", var"#s4"<:(CategoricalArrays.CategoricalArray{T, var"#s3", R, V, C, U} where {var"#s3", V, C, U}), I, L}, CategoricalArrays.CategoricalArray{T, N, R, V, C, U} where {V, C, U}} where {T, N, R<:Integer}, src::AbstractArray) in CategoricalArrays at /Users/anthony/.julia/packages/CategoricalArrays/rDwMt/src/array.jl:615
  copyto!(dest::AbstractVector{T} where T, src::ChainedVector) in SentinelArrays at /Users/anthony/.julia/packages/SentinelArrays/1hMOA/src/chainedvector.jl:326
Possible fix, define
  copyto!(::Union{SubArray{T, 1, var"#s4", I, L} where {T, T1, R<:Integer, var"#s4"<:(CategoricalArrays.CategoricalArray{T1, var"#s3", R, V, C, U} where {var"#s3", V, C, U}), I, L}, CategoricalArrays.CategoricalVector{T, R, V, C, U} where {C, U, T, R<:Integer, V}}, ::ChainedVector)                                                                              
Stacktrace:
 [1] _convert(::Type{CategoricalArrays.CategoricalVector{Int64, UInt32, V, C, U} where {V, C, U}}, A::ChainedVector{Int64, UnitRange{Int64}}; levels::Nothing)
   @ CategoricalArrays ~/.julia/packages/CategoricalArrays/rDwMt/src/array.jl:343
 [2] (CategoricalArrays.CategoricalVector{Int64, UInt32, V, C, U} where {V, C, U})(A::ChainedVector{Int64, UnitRange{Int64}}; levels::Nothing, ordered::Bool)                     
   @ CategoricalArrays ~/.julia/packages/CategoricalArrays/rDwMt/src/array.jl:253
 [3] #categorical#72
   @ ~/.julia/packages/CategoricalArrays/rDwMt/src/array.jl:927 [inlined]
 [4] categorical(A::ChainedVector{Int64, UnitRange{Int64}})
   @ CategoricalArrays ~/.julia/packages/CategoricalArrays/rDwMt/src/array.jl:926
 [5] top-level scope
   @ REPL[83]:1
```

```julia
  [324d7699] CategoricalArrays v0.10.0
  [91c51154] SentinelArrays v1.3.5
```

```julia
julia> versioninfo()
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.6.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_LTS_PATH = /Applications/Julia-1.0.app/Contents/Resources/julia/bin/julia
  JULIA_PATH = /Applications/Julia-1.6.app/Contents/Resources/julia/bin/julia
  JULIA_EGLOT_PATH = /Applications/Julia-1.5.app/Contents/Resources/julia/bin/julia
  JULIA_NUM_THREADS = 5
  JULIA_NIGHTLY_PATH = /Applications/Julia-1.7.app/Contents/Resources/julia/bin/julia
```

### PR DESCRIPTION
**Pull request by nalimilan**
Fix #361 adding special methods for `ChainedVector`, and make as many dependencies as possible optional using Requires.jl.

`@time using CategoricalArrays` on a fresh session goes down from 0.26s to 0.18s.

### CODE REVIEW

File: src/CategoricalArrays.jl [hunk 1]
```diff
@@ -36,4 +34,74 @@ module CategoricalArrays
...
+
+    function __init__()
+        @require JSON="682c06a0-de6a-54ab-a142-c8b1cf79cde6" begin
+            #import JSON
+            # JSON of CategoricalValue is JSON of the value it refers to
+            JSON.lower(x::CategoricalValue) = JSON.lower(unwrap(x))
+            DataAPI.defaultarray(::Type{CategoricalValue{T, R}}, N) where {T, R} =
+                CategoricalArray{T, N, R}
+            DataAPI.defaultarray(::Type{Union{CategoricalValue{T, R}, Missing}}, N) where {T, R} =
+                CategoricalArray{Union{T, Missing}, N, R}
```
Comment (bkamins) [inline src/CategoricalArrays.jl:46]: why are these definitions linked to JSON? I would assume they make sense in general.

File: src/CategoricalArrays.jl [hunk 2]
```diff
@@ -36,4 +34,74 @@ module CategoricalArrays
...
+        end
+
+        @require RecipesBase="3cdcf5f2-1ef4-517c-9805-6587b60abb01" @eval begin
```
Comment (bkamins) [inline src/CategoricalArrays.jl:49]: it would be good to cover this with tests if it is not overly complex.

File: src/CategoricalArrays.jl [hunk 3]
```diff
@@ -36,4 +34,74 @@ module CategoricalArrays
...
+        end
+
+        @require SentinelArrays="91c51154-3ec4-41a3-a24f-3f23e20d615c" begin
+            #import SentinelArrays
+            copyto!(dest::CatArrOrSub{<:Any, 1}, src::SentinelArrays.ChainedVector) =
+            copyto!(dest, 1, src, 1, length(src))
+            copyto!(dest::CatArrOrSub{<:Any, 1}, dstart::Union{Signed, Unsigned},
+                    src::SentinelArrays.ChainedVector, sstart::Union{Signed, Unsigned},
+                    n::Union{Signed, Unsigned}) =
+            invoke(copyto!, Tuple{AbstractArray, Union{Signed, Unsigned},
```
Comment (bkamins) [inline src/CategoricalArrays.jl:65]: indent

**Comment by bkamins** [comment]
I understand that the proper code blocks will be executed both when:
1. CategoricalArrays.jl being loaded and the given package was already loaded directly or indirectly
2. CategoricalArrays.jl was loaded and the given package is being loaded directly or indirectly

Right?

File: src/CategoricalArrays.jl (hunk 1 above)
Comment (nalimilan) [reply src/CategoricalArrays.jl:46]: Woops. Unfortunately tests don't catch this as JSON is loaded earlier...

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (nalimilan) [reply src/CategoricalArrays.jl:49]: It's tested AFAICT. Or do you mean something else?

**Comment by nalimilan** [comment]
Yes, at least that's my understanding. :-)

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (bkamins) [reply src/CategoricalArrays.jl:44]: Just coveage test was signaling it is not tested. That is why I asked.

File: Project.toml [hunk 4]
```diff
@@ -5,12 +5,10 @@ version = "0.10.0"
 [deps]
```
Comment (bkamins) [inline Project.toml:5]: maybe bump package version for a release?

**Review by bkamins — APPROVED** [review]

File: Project.toml (hunk 4 above)
Comment (nalimilan) [reply Project.toml:5]: I'll do that directly on master after merging.

File: src/CategoricalArrays.jl (hunk 2 above)
Comment (nalimilan) [reply src/CategoricalArrays.jl:44]: You mean line 50 at https://github.com/JuliaData/CategoricalArrays.jl/runs/3633298660? That's probably a bug, as other lines in the block appear to be covered. I guess a macro within two macro calls is too complex for coverage checks.

Comment (bkamins) [reply src/CategoricalArrays.jl:44]: ok

**Comment by nalimilan** [comment]
https://github.com/JuliaRegistries/General/pull/45088

**Pull request merged**